File: 20.valid-parentheses.cs
using System;
using System.Collections.Generic;

public class Solution20
{
    public bool IsValid(string s) 
    {
        if (String.IsNullOrEmpty(s))
        {
            return true;
        }
        
        // For this kind of question related to parentheses, first choice is to try stack
        // When using stack, there are three logics
        // Push
        // Peek
        // Pop
        // Also use .Count to check if the stack is empty or not.
        Stack<char> stack = new Stack<char>();
        foreach (var c in s)
        {
            if (stack.Count > 0 &&
                (stack.Peek() == '(' && c == ')' ||
                stack.Peek() == '{' && c == '}' ||
                stack.Peek() == '[' && c == ']'))
            {
                stack.Pop();
            }
            else
            {
                stack.Push(c);
            }
        }

        return stack.Count == 0;
    }
}

File: 20_valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        records = []
        if len(s) == 0:
            return True
        
        dict = {')':'(', ']':'[', '}':'{'}
        for i in range(len(s)):
            if s[i] not in dict:
                records.append(s[i])
            elif len(records)> 0 and dict[s[i]]== records[len(records)-1]:
                records.pop()
            else:
                records.append(s[i])
        
        if len(records) ==0:
            return True

        return False

instance = Solution()
print(instance.isValid("()"))


